### ISSUE TITLE
Intermittent 404 error when deployed in Tomcat 7 

### ISSUE THREAD

**Issue by sergiomichels** [grails/grails-core#9901]
### Steps to Reproduce

Please fork my [sample application](https://github.com/sergiomichels/grails3-404-issue/tree/master/teste2)
 
1. Create a prod war
2. Deploy it in one Tomcat 7 (I'm using apache-tomcat-7.0.55)
3. Try to access http://localhost:8080/teste2/teste/index. Sometimes it will display the page and others will show the /error page. To test the intermittent behavior just restart and redeploy the **same war**. 

### Expected Behaviour

The expected in this scenario is to always show the page, since it's not secured. I'm not confident if this is a Grails core or Spring Security plugin bug, so I'm reporting it here.

### Actual Behaviour

Sometimes the flow will end in the /error page, and others will display it correctly.

### Environment Information

- **Operating System**: Windows 8.1 or 10
- **Grails Version**: 3.1.4
- **Groovy Version**: 2.4.6
- **JVM Version**: 1.7.0_79
- **Tomcat Version**: 7.0.55

### Debug information

I nailed down to `UrlMappingsHandlerMapping` class, you can see that I created a new class to debug it. I can notice that sometimes it handles correctly and return the `org.grails.web.mapping.mvc.GrailsControllerUrlMappingInfo` intance, but sometimes the flow will use one `org.grails.web.mapping.DefaultUrlMappingInfo` instance, causing the 404 error. Maybe the problem is inside the `urlMappingsHolder` class.

Here's the debug that will be printed when the flow ends incorrectly

```
DEBUG org.grails.web.mapping.DefaultUrlMappingsHolder - Attempting to match URI
[/teste/index] with pattern [/oauth/(*)/(*)]
DEBUG org.grails.web.mapping.DefaultUrlMappingsHolder - Attempting to match URI
[/teste/index] with pattern [/(*)/(*)?/(*)?(.(*))?]
DEBUG org.grails.web.mapping.DefaultUrlMappingsHolder - Matched URI [/teste/inde
x] with pattern [/(*)/(*)?/(*)?(.(*))?], adding to posibilities
INFO org.grails.web.mapping.mvc.CwiUrlMappingsHandler - LOOP in UrlMappingsInfo
availables for the match: org.grails.web.mapping.mvc.GrailsControllerUrlMappings
@6399bced
INFO org.grails.web.mapping.mvc.CwiUrlMappingsHandler - UrlMappingInfo: /(*)/(*)
?/(*)?(.(*))? , class org.grails.web.mapping.DefaultUrlMappingInfo
INFO org.grails.web.mapping.mvc.CwiUrlMappingsHandler - RETURN NULL!
INFO org.grails.web.mapping.mvc.CwiUrlMappingsHandler - getHandlerInternal() : n
ull
```

**Any report is really appreciated, since this error is blocking us** from deploying this application in the production environment (run-app works).

**Comment by graemerocher** [comment]
When you "restart and redeploy" do you just restart Tomcat? What do you mean by redeploy? Are you copying a new version of the JAR when Tomcat is running? 

So far I have restarted Tomcat 10 times and have not been able to reproduce

**Comment by graemerocher** [comment]
The only thing I can think would cause this is some kind of concurrency bug in the JDK version you are using. Although even then I don't see how. However, a starting point for you would be to attach a debugger and see what happens in this method 

https://github.com/grails/grails-core/blob/master/grails-web-url-mappings/src/main/groovy/org/grails/web/mapping/mvc/AbstractGrailsControllerUrlMappings.groovy#L128

This is probably a duplicate of #9676 however, we have yet to be able to reproduce it. I will try again with your example in Windows shortly

**Comment by graemerocher** [comment]
Tested on Windows 7 again with the following setup:

* *Grails Version* - 3.1.5
* *Tomcat Version* - 7.0.68
* *JDK Version* - 1.7.0_55

Still not able to reproduce. I advise that you navigate to your Tomcat `bin` directory and start Tomcat using (assuming Windows):

        java -Xdebug -Xrunjdwp:transport=dt_socket,server=y,suspend=y,address=5005 -cp "bootstrap.jar;tomcat-juli.jar" org.apache.catalina.startup.Bootstrap

Attach a remote debugger and set a break point around here:

https://github.com/grails/grails-core/blob/master/grails-web-url-mappings/src/main/groovy/org/grails/web/mapping/mvc/AbstractGrailsControllerUrlMappings.groovy#L52

And observe whether the controllers are being registered correctly. My suspicion is this is a JDK specific concurrency bug, but without the ability to reproduce we would need feedback from yourself to advance this issue.

**Comment by sergiomichels** [comment]
By redeploy I mean that I restart Tomcat and let him redeploy the same war that's in the webapps folder.

I will debug it and reply later, thanks!

**Comment by sergiomichels** [comment]
The only combination that works for me is the latest Tomcat with 1.7.0_55. I tried

Fail
 - Tomcat Version: 7.0.55
 - JDK Version - 1.7.0_55

 - Tomcat Version: 7.0.69
 - JDK Version - 1.7.0_79

Can you try with the reported JDK?

**Comment by graemerocher** [comment]
Will try...

**Comment by sergiomichels** [comment]
With the debugger I can see that Grails is creating **two instances** of the UrlMappings. The first instance doesn't have the controller and the second one have.

I can see also that console log two times the message "Configuring Spring Security".

IDEA Community have the option to export frames?

**Comment by graemerocher** [comment]
That sounds like there are two `ApplicationLoader` instances. Can you place  a break point on 

https://github.com/grails/grails-core/blob/3.1.x/grails-web-boot/src/main/groovy/org/grails/boot/context/web/GrailsAppServletInitializer.groovy#L34

And then post screenshots of the call stack each time the breakpoint is hit.

Thanks.

**Comment by sergiomichels** [comment]
![grails-stack-one](https://cloud.githubusercontent.com/assets/1640822/14859997/0c156ea4-0c7d-11e6-9c2e-131ce924cf88.PNG)
![grails-stack-two](https://cloud.githubusercontent.com/assets/1640822/14859996/0c12cc6c-0c7d-11e6-8811-5820e0ee951b.PNG)

**Comment by sergiomichels** [comment]
Sorry, it's two application loaders.

![grails-apploader-1](https://cloud.githubusercontent.com/assets/1640822/14860144/c1c26054-0c7d-11e6-9d5a-3331270918a2.PNG)
![grails-apploader-2](https://cloud.githubusercontent.com/assets/1640822/14860143/c1c043aa-0c7d-11e6-8ebc-a851ea65c129.PNG)

**Comment by graemerocher** [comment]
Thanks, can you find out in the debugger the full class names of those 2 loaders? This feedback is valuable thanks.

**Comment by sergiomichels** [comment]
It's `export.ApplicationLoader` and `teste2.ApplicationLoader`.

**Comment by sergiomichels** [comment]
Looking at the plugin, it exposes
 - One test controller
 - One custom converter

I will ask them to try load after controllers, but I'm not sure if this is the issue.

**Comment by sergiomichels** [comment]
Even with `loadAfter` the problem still happens. The plugin is now 0.0.4

**Comment by sergiomichels** [comment]
The application works when the first loader handled is the application one.

**Comment by graemerocher** [comment]
This is a problem with the build of your `export` plugin in that case. An `ApplicationLoader` is only generated when building a plugin: 

https://github.com/grails/grails-core/blob/master/grails-web-boot/src/main/groovy/org/grails/compiler/boot/BootInitializerClassInjector.groovy#L57

Check your `build.gradle` of the `export` plugin and make sure the `grails-plugin` Gradle plugin has been applied

**Comment by sergiomichels** [comment]
You mean this?

```
version "0.0.3"
group "export"

apply plugin:"com.jfrog.artifactory"
apply plugin:"eclipse"
apply plugin:"idea"
apply plugin:"org.grails.grails-plugin"
apply plugin:"org.grails.grails-plugin-publish"
apply plugin:"org.grails.grails-gsp"
```

**Comment by sergiomichels** [comment]
I'm pretty sure it was created by the `create-plugin` command.

**Comment by graemerocher** [comment]
Try a `gradle clean install` on your plugin project, I suspect the issue is that IntelliJ compiled classes were used when building your plugin

**Comment by sergiomichels** [comment]
Huge thanks! Now all makes sense, and my test application works correctly.

**Comment by graemerocher** [comment]
Fixed by #9904
